### PR TITLE
Introduce ServiceLoader use for TargetClients

### PR DESCRIPTION
**Pull request by lmccay**
#304 represents the intent of this PR towards the larger extensibility improvements.

## What is the purpose of the pull request

This change will introduce the use of ServiceLoader by the TableFormatClientFactory to discover the TargetClient implementations by TableFormat which was recently changed from an Enum to a String in preparation for this change.

ServiceLoader will load all implementations on the classpath (built-in to OneTable or added by the user) and select the intended client implementation by name provided for the Sync via argument or configuration.

This will allow users and/or tooling vendors to extend the capabilities without requiring all TargetClient's to be committed to the project.

## Brief change log

* Refactored TableFormatClientFactory to use ServiceLoader
* Refactored existing clients to have a default no-arg ctor and an init() for initializing the client with required config
* Removed final qualifiers from members so that init() can set them after construction
* Extracted interfaces from PerTableConfig, HudiSourceConfig, IcebergCatalogConfig in order to maintain the decoupling of the api and core modules after needing the config files in the init()
* Moved to the use of the PerTableConfig interface wherever possible

## Verify this pull request

This pull request is already covered by many existing tests.
Also added a new test in TestTableFormatClientFactory for the factory use in general but also specifically for the use of ServiceLoader.

### CODE REVIEW

**Comment by lmccay** [comment]
Happy New Year PR for onetable, @the-other-tim-brown

File: core/src/main/java/io/onetable/client/TableFormatClientFactory.java [hunk 1]
```diff
@@ -43,15 +38,21 @@ public static TableFormatClientFactory getInstance() {
...
-        throw new NotSupportedException("Target format is not yet supported: " + tableFormat);
+    TargetClient targetClient = createTargetClientForName(tableFormat);
+    if (targetClient == null) {
+      throw new NotSupportedException("Target format is not yet supported: " + tableFormat);
     }
+    targetClient.init(perTableConfig, configuration);
+    return targetClient;
+  }
+
+  public TargetClient createTargetClientForName(String tableFormatName) {
```
Comment (ashvina) [inline core/src/main/java/io/onetable/client/TableFormatClientFactory.java:49]: nit: This method does not require public access.

File: core/src/main/java/io/onetable/client/TableFormatClientFactory.java [hunk 2]
```diff
@@ -43,15 +38,21 @@ public static TableFormatClientFactory getInstance() {
...
-      case ICEBERG:
-        return new IcebergClient(perTableConfig, configuration);
-      case DELTA:
-        return new DeltaClient(perTableConfig, configuration);
-      case HUDI:
-        return new HudiTargetClient(perTableConfig, configuration);
-      default:
-        throw new NotSupportedException("Target format is not yet supported: " + tableFormat);
+    TargetClient targetClient = createTargetClientForName(tableFormat);
+    if (targetClient == null) {
```
Comment (ashvina) [inline core/src/main/java/io/onetable/client/TableFormatClientFactory.java:42]: Is it possible for `targetClient` to have a null value here?

File: api/src/main/java/io/onetable/client/PerTableConfig.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * limitations under the License.
+ */
+ 
+package io.onetable.client;
+
+import java.util.List;
+
+import io.onetable.model.sync.SyncMode;
+
+public interface PerTableConfig {
```
Comment (ashvina) [inline api/src/main/java/io/onetable/client/PerTableConfig.java:25]: This interface appears to have little value. I don't think we need to change `PerTableConfig` for `ServiceLoader` integration. We can use the existing class to initialize the target clients for now.

File: core/src/main/java/io/onetable/client/PerTableConfigImpl.java [hunk 4]
```diff
@@ -30,13 +30,13 @@
 
 import com.google.common.base.Preconditions;
 
-import io.onetable.hudi.HudiSourceConfig;
+import io.onetable.hudi.HudiSourceConfigImpl;
 import io.onetable.iceberg.IcebergCatalogConfig;
 import io.onetable.model.sync.SyncMode;
 
 /** Represents input configuration to the sync process. */
 @Value
```
Comment (ashvina) [inline core/src/main/java/io/onetable/client/PerTableConfigImpl.java:38]: There is a conflict here, this class is an immutable entity class that uses lombok and implements the new interface.

File: core/src/main/java/io/onetable/client/TableFormatClientFactory.java (hunk 1 above)
Comment (lmccay) [reply core/src/main/java/io/onetable/client/TableFormatClientFactory.java:49]: This depends on the programming model going forward. This allows for application code to create a targetClient then set values and call init(). If we want to completely hide that going forward then we can make this private. I figured we would let that be decided as a follow up but can change it now and we can always open it up later.

File: core/src/main/java/io/onetable/client/TableFormatClientFactory.java (hunk 2 above)
Comment (lmccay) [reply core/src/main/java/io/onetable/client/TableFormatClientFactory.java:42]: Yes. If there are no implementations that match the provided format name from those loaded by ServiceLoader.

File: api/src/main/java/io/onetable/client/PerTableConfig.java (hunk 3 above)
Comment (lmccay) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:25]: This was an open question in the filed issue. We either need to extract an interface, add a dependency from api on core or move PerTableConfig into api. If we are going to move to an injection model where the config doesn't need to be known by the clients - which I would suggest we do - then it can go completely away and even the use of the interface within init() could be removed. All of the interfaces that were extracted in this PR are in order to avoid moving implementation classes or adding dependencies to api module on core.

File: core/src/main/java/io/onetable/iceberg/IcebergClient.java [hunk 5]
```diff
@@ -94,6 +96,26 @@ public IcebergClient(PerTableConfig perTableConfig, Configuration configuration)
       IcebergPartitionSpecSync partitionSpecSync,
       IcebergDataFileUpdatesSync dataFileUpdatesExtractor,
       IcebergTableManager tableManager) {
+    _init(
```
Comment (ashvina) [inline core/src/main/java/io/onetable/iceberg/IcebergClient.java:99]: This constructor should only be used for testing purposes and should have an appropriate annotation.

File: core/src/main/java/io/onetable/client/PerTableConfigImpl.java (hunk 4 above)
Comment (lmccay) [reply core/src/main/java/io/onetable/client/PerTableConfigImpl.java:38]: This, I don't follow. I am only just learning about lombok now and need more information. There is no compile or runtime conflict here unless there is a lack of test coverage.

File: core/src/main/java/io/onetable/hudi/HudiTargetClient.java [hunk 6]
```diff
@@ -96,18 +96,20 @@
...
+  private CommitStateCreator commitStateCreator;
+  private int timelineRetentionInHours;
+  private int maxNumDeltaCommitsBeforeCompaction;
+  private String tableDataPath;
   private Optional<HoodieTableMetaClient> metaClient;
   private CommitState commitState;
 
+  public HudiTargetClient() {}
+
   public HudiTargetClient(PerTableConfig perTableConfig, Configuration configuration) {
```
[the remark this answers is not in the document]
Comment (lmccay) [reply core/src/main/java/io/onetable/hudi/HudiTargetClient.java:111]: If we can assume that there are no external uses that directly instantiate the class then we can remove it. This is an backward compatible refactoring which is what I normally do. Should we just remove or deprecate first?

File: core/src/main/java/io/onetable/delta/DeltaClient.java [hunk 7]
```diff
@@ -99,6 +101,25 @@ public DeltaClient(PerTableConfig perTableConfig, SparkSession sparkSession) {
       DeltaSchemaExtractor schemaExtractor,
       DeltaPartitionExtractor partitionExtractor,
       DeltaDataFileUpdatesExtractor dataFileUpdatesExtractor) {
+
+    _init(
```
[the remark this answers is not in the document]
Comment (lmccay) [reply core/src/main/java/io/onetable/delta/DeltaClient.java:105]: package protected ctor's are usually used for testing but I guess could be used by extensions as well. If we can assume it is not used externally then we can deprecate or remove. Again, I generally refactor in a backwards compatible way but will do what you like here.

File: core/src/main/java/io/onetable/iceberg/IcebergClient.java [hunk 8]
```diff
@@ -56,21 +56,23 @@
...
+  private IcebergCatalogConfig catalogConfig;
+  private Configuration configuration;
+  private int snapshotRetentionInHours;
   private Transaction transaction;
   private Table table;
   private OneTable internalTableState;
 
+  public IcebergClient() {}
+
   public IcebergClient(PerTableConfig perTableConfig, Configuration configuration) {
```
[the remark this answers is not in the document]
Comment (lmccay) [reply core/src/main/java/io/onetable/iceberg/IcebergClient.java:76]: Same comment as other ctor responses. We can deprecate or remove. Let me know which you'd  prefer.

File: core/src/main/java/io/onetable/iceberg/IcebergClient.java (hunk 5 above)
Comment (lmccay) [reply core/src/main/java/io/onetable/iceberg/IcebergClient.java:99]: That is outside the scope of the refactoring here, if wasn't there previously. If it wasn't there then it may be used externally and this would make this PR backward incompatible unnecessarily. Maybe we can make that a separate change?

File: api/src/main/java/io/onetable/client/PerTableConfig.java (hunk 3 above)
Comment (ashvina) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:25]: IMO, moving `PerTableConfig` to api is cleaner than creating this interface. This also means that `HudiSourceConfig` and `IcebergCatalog` will also need to be moved to `api`.  To me this is better than creating the interfaces. I had created another issue to clean up the per table config, which could be used to track cleaning of these classes also. 

I am also supportive of dependency injection. Which framework did you have in mind?

File: core/src/main/java/io/onetable/hudi/HudiTargetClient.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ashvina) [reply core/src/main/java/io/onetable/hudi/HudiTargetClient.java:111]: I think this was added for testing and should have been annotated. I prefer removing it.

File: core/src/main/java/io/onetable/client/TableFormatClientFactory.java (hunk 1 above)
Comment (ashvina) [reply core/src/main/java/io/onetable/client/TableFormatClientFactory.java:49]: If this method is meant for public use, please provide a javadoc comment that explains its purpose.

File: core/src/main/java/io/onetable/client/TableFormatClientFactory.java (hunk 2 above)
Comment (ashvina) [reply core/src/main/java/io/onetable/client/TableFormatClientFactory.java:42]: Would line 56 throw an exception if no match is found?

Comment (lmccay) [reply core/src/main/java/io/onetable/client/TableFormatClientFactory.java:42]: Yep - you're right. I was thinking about the code before you asked me to not return null. :)

File: core/src/main/java/io/onetable/client/TableFormatClientFactory.java (hunk 1 above)
Comment (lmccay) [reply core/src/main/java/io/onetable/client/TableFormatClientFactory.java:49]: I guess that I'll add one to the existing method as well...

File: api/src/main/java/io/onetable/client/PerTableConfig.java (hunk 3 above)
Comment (lmccay) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:25]: Interfaces are definitely cleaner for testing and api/spi modules but we can certainly move it. It shouldn't pull in other dependencies since I already extracted the other interfaces. I'll give it a shot.

File: core/src/main/java/io/onetable/hudi/HudiTargetClient.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (lmccay) [reply core/src/main/java/io/onetable/hudi/HudiTargetClient.java:111]: Will do.

File: core/src/main/java/io/onetable/iceberg/IcebergClient.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (lmccay) [reply core/src/main/java/io/onetable/iceberg/IcebergClient.java:76]: Since you said you prefer to remove it from HudiTargetClient, I'll remove this as well.

File: core/src/main/java/io/onetable/delta/DeltaClient.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (lmccay) [reply core/src/main/java/io/onetable/delta/DeltaClient.java:105]: done.

File: core/src/main/java/io/onetable/hudi/HudiTargetClient.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (lmccay) [reply core/src/main/java/io/onetable/hudi/HudiTargetClient.java:111]: done.

File: core/src/main/java/io/onetable/iceberg/IcebergClient.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (lmccay) [reply core/src/main/java/io/onetable/iceberg/IcebergClient.java:76]: done.

File: core/src/main/java/io/onetable/iceberg/IcebergClient.java (hunk 5 above)
Comment (lmccay) [reply core/src/main/java/io/onetable/iceberg/IcebergClient.java:99]: Done. We can always easily change it back if we need to.

File: core/src/main/java/io/onetable/client/TableFormatClientFactory.java (hunk 2 above)
Comment (lmccay) [reply core/src/main/java/io/onetable/client/TableFormatClientFactory.java:42]: done.

File: core/src/main/java/io/onetable/client/TableFormatClientFactory.java (hunk 1 above)
Comment (lmccay) [reply core/src/main/java/io/onetable/client/TableFormatClientFactory.java:49]: done.

File: api/src/main/java/io/onetable/client/PerTableConfig.java (hunk 3 above)
Comment (lmccay) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:25]: @ashvina - I don't have strong opinions about dependency injection frameworks here. I think rolling our own for this simple case would work just fine. If we moved to a more generic name-value pair config that would make it really easy.

I've tried to move the PerTableConfigImpl into api and like you said it will pull the Target specific configs along with it which is kinda messy. Do you want them in target specific sub-packages or just in the client package?

Personally, I think that the interface based approach that I did here to get ServiceLoader working is sufficient for this PR and the config refactoring could follow up and clean that up however appropriate. Mixing them in this PR complicates this work and will still need cleanup in the follow on anyway.

Thoughts on that?

File: core/src/main/java/io/onetable/client/PerTableConfigImpl.java (hunk 4 above)
Comment (lmccay) [reply core/src/main/java/io/onetable/client/PerTableConfigImpl.java:38]: @ashvina - can you give me more information on what I should do to address this?

**Comment by lmccay** [comment]
Resolved most of the review comments, @ashvina .
I've left the two open topics open with comments/questions.

File: api/src/main/java/io/onetable/client/PerTableConfig.java (hunk 3 above)
Comment (the-other-tim-brown) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:25]: I don't like the idea of pulling the hudi and iceberg specifics into the api module. If we can avoid that, then that is preferred. We may need to do a larger refactor to do this though so I don't want to block progress for something that will probably require its own PR. I think we can try to pull each format's specifics into their own module in a future refactor and that can help us really find what is common and may help us structure this better.

File: core/src/main/java/io/onetable/client/PerTableConfigImpl.java (hunk 4 above)
Comment (the-other-tim-brown) [reply core/src/main/java/io/onetable/client/PerTableConfigImpl.java:38]: Lombok's `@Value` annotation will generate some code for you. Mainly it marks all fields as `private final` and adds getters for you https://projectlombok.org/features/Value

I am not sure I see why this is a conflict though @ashvina since there are no setters on the interface

File: api/src/main/java/io/onetable/client/PerTableConfig.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+  HudiSourceConfig getHudiSourceConfig();
+
+  List<String> getTargetTableFormats();
+
+  SyncMode getSyncMode();
+
+  String[] getNamespace();
+
+  CatalogConfig getIcebergCatalogConfig();
```
Comment (the-other-tim-brown) [inline api/src/main/java/io/onetable/client/PerTableConfig.java:42]: Do we want this to just be `getCatalogConfig` and then have an Iceberg specific one? It could be possible to have a hive metastore catalog or something like that for Delta/Hudi as well

File: api/src/main/java/io/onetable/client/PerTableConfig.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+public interface PerTableConfig {
+  int getTargetMetadataRetentionInHours();
+
+  String getTableBasePath();
+
+  String getTableDataPath();
+
+  String getTableName();
+
+  HudiSourceConfig getHudiSourceConfig();
```
Comment (the-other-tim-brown) [inline api/src/main/java/io/onetable/client/PerTableConfig.java:34]: Do we want to split this out into a separate source config interface that each source client can implement if it needs anything special?

File: core/src/main/java/io/onetable/hudi/HudiSourceConfigImpl.java [hunk 11]
```diff
@@ -47,12 +48,20 @@ public HudiSourceConfig(String partitionSpecExtractorClass, String partitionFiel
...
 
   @Value
-  public static class PartitionFieldSpec {
+  static class PartitionFieldSpec {
     String sourceFieldPath;
     PartitionTransformType transformType;
     String format;
   }
 
+  public String getPartitionSpecExtractorClass() {
```
Comment (the-other-tim-brown) [inline core/src/main/java/io/onetable/hudi/HudiSourceConfigImpl.java:57]: These getters are redundant with lombok, right?

File: core/src/main/java/io/onetable/iceberg/IcebergCatalogConfig.java [hunk 12]
```diff
@@ -27,8 +27,20 @@
 
 @Value
 @Builder
-public class IcebergCatalogConfig {
+public class IcebergCatalogConfig implements io.onetable.client.CatalogConfig {
   @NonNull String catalogImpl;
   @NonNull String catalogName;
   @NonNull @Builder.Default Map<String, String> catalogOptions = Collections.emptyMap();
+
+  public String getCatalogImpl() {
```
Comment (the-other-tim-brown) [inline core/src/main/java/io/onetable/iceberg/IcebergCatalogConfig.java:35]: Similar nitpick on the getters here, lombok is already adding these

File: core/src/test/java/io/onetable/client/TestTableFormatClientFactory.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+    conf.setStrings("spark.master", "local");
+    tc.init(perTableConfig, conf);
+    assertNotNull(tc.getTableFormat().equalsIgnoreCase(TableFormat.ICEBERG));
+  }
+
+  @Test
+  public void testTableClientFromNameForUNKOWN() {
+    try {
+      TargetClient tc = TableFormatClientFactory.getInstance().createTargetClientForName("UNKOWN");
+      fail("NotSupportedException expected and operation succeeded inappropriately.");
```
Comment (the-other-tim-brown) [inline core/src/test/java/io/onetable/client/TestTableFormatClientFactory.java:81]: There is an `assertThrows` that you can use to test this

File: core/src/test/java/io/onetable/client/TestTableFormatClientFactory.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+  public void testTableClientFromNameForDELTA() {
+    TargetClient tc =
+        TableFormatClientFactory.getInstance().createTargetClientForName(TableFormat.DELTA);
+    assertNotNull(tc);
+    PerTableConfig perTableConfig =
+        getPerTableConfig(Arrays.asList(TableFormat.DELTA), SyncMode.INCREMENTAL);
+    Configuration conf = new Configuration();
+    conf.setStrings("spark.master", "local");
+    tc.init(perTableConfig, conf);
+    assertNotNull(tc.getTableFormat().equalsIgnoreCase(TableFormat.DELTA));
```
Comment (the-other-tim-brown) [inline core/src/test/java/io/onetable/client/TestTableFormatClientFactory.java:48]: Can you use `assertEquals` for this and the other cases below so it is more clear?

File: core/src/main/java/io/onetable/client/PerTableConfigImpl.java (hunk 4 above)
Comment (the-other-tim-brown) [reply core/src/main/java/io/onetable/client/PerTableConfigImpl.java:38]: All of the getters in this class are being generated by lombok so let's just rely on those instead of also implementing them?

**Review by the-other-tim-brown — COMMENTED** [review]
Much easier to review this without those string changes :) Thanks for breaking this up for us

File: core/src/main/java/io/onetable/client/PerTableConfigImpl.java (hunk 4 above)
Comment (lmccay) [reply core/src/main/java/io/onetable/client/PerTableConfigImpl.java:38]: hmmm, I'd expect that I implemented them due to red lines in my IDE. I'll double check.

File: api/src/main/java/io/onetable/client/PerTableConfig.java (hunk 9 above)
Comment (lmccay) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:42]: @ashvina is planning a PerTableConfig refactor. I figured it better to leave these until then rather than change something that may need to be changed again anyway. What do you think?

File: api/src/main/java/io/onetable/client/PerTableConfig.java (hunk 10 above)
Comment (lmccay) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:34]: Same as above.

File: core/src/main/java/io/onetable/client/PerTableConfigImpl.java (hunk 4 above)
Comment (lmccay) [reply core/src/main/java/io/onetable/client/PerTableConfigImpl.java:38]: done.

File: core/src/main/java/io/onetable/hudi/HudiSourceConfigImpl.java (hunk 11 above)
Comment (lmccay) [reply core/src/main/java/io/onetable/hudi/HudiSourceConfigImpl.java:57]: done.

File: core/src/main/java/io/onetable/iceberg/IcebergCatalogConfig.java (hunk 12 above)
Comment (lmccay) [reply core/src/main/java/io/onetable/iceberg/IcebergCatalogConfig.java:35]: done.

File: core/src/main/java/io/onetable/delta/DeltaClient.java [hunk 15]
```diff
@@ -67,15 +69,17 @@ public class DeltaClient implements TargetClient {
...
 
-  private final String tableName;
-  private final int logRetentionInHours;
+  private String tableName;
+  private int logRetentionInHours;
   private TransactionState transactionState;
 
+  public DeltaClient() {}
+
   public DeltaClient(PerTableConfig perTableConfig, Configuration configuration) {
```
[the remark this answers is not in the document]
Comment (lmccay) [reply core/src/main/java/io/onetable/delta/DeltaClient.java:83]: done.

File: core/src/test/java/io/onetable/client/TestTableFormatClientFactory.java (hunk 13 above)
Comment (lmccay) [reply core/src/test/java/io/onetable/client/TestTableFormatClientFactory.java:81]: done.

File: core/src/test/java/io/onetable/client/TestTableFormatClientFactory.java (hunk 14 above)
Comment (lmccay) [reply core/src/test/java/io/onetable/client/TestTableFormatClientFactory.java:48]: Yikes! :)

Comment (lmccay) [reply core/src/test/java/io/onetable/client/TestTableFormatClientFactory.java:48]: done.

**Comment by lmccay** [comment]
@the-other-tim-brown , @ashvina - addressed those issues that were clear. Left a couple open for resolution of questions.

File: core/src/main/java/io/onetable/client/PerTableConfigImpl.java (hunk 4 above)
Comment (the-other-tim-brown) [reply core/src/main/java/io/onetable/client/PerTableConfigImpl.java:38]: There is a lombok plugin for intellij that you may need to install

File: core/src/test/java/io/onetable/client/TestTableFormatClientFactory.java (hunk 14 above)
Comment (the-other-tim-brown) [reply core/src/test/java/io/onetable/client/TestTableFormatClientFactory.java]: Curious if this was GitHub co-pilot's recommended code. It has given me some interesting autocompletes on tests haha

File: api/src/main/java/io/onetable/client/PerTableConfig.java (hunk 9 above)
Comment (the-other-tim-brown) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:42]: Ok I'll defer to @ashvina on this one. I am in favor of smaller change sets in general so makes sense to do that on its own when thinking about it that way.

File: core/src/test/java/io/onetable/client/TestTableFormatClientFactory.java (hunk 14 above)
Comment (lmccay) [reply core/src/test/java/io/onetable/client/TestTableFormatClientFactory.java]: Nope - just left over from when I was previously just making sure it was created. Forgot to change the assert. :)

File: api/src/main/java/io/onetable/client/PerTableConfig.java (hunk 3 above)
Comment (ashvina) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:25]: Yes, format specific in `api` is not a good design. The introduction of an interface for `PerTableConfig`, which also had format specific contract, did not address this design issue. Fixing this requires refactoring. I have started a discussion thread towards this #297

File: core/src/main/java/io/onetable/client/PerTableConfigImpl.java (hunk 4 above)
Comment (ashvina) [reply core/src/main/java/io/onetable/client/PerTableConfigImpl.java:38]: @the-other-tim-brown , the `conflict` / duplication was in the getters, they were generated by Lombok and were required by the interface as well.

File: api/src/main/java/io/onetable/client/PerTableConfig.java (hunk 9 above)
Comment (ashvina) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:42]: 👍  to addressing one issue in one PR.
A good starting point would be to finalize the design of `PerTableConfig` (#297) and then proceed accordingly.

Comment (lmccay) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:42]: @ashvina - I read that as agreeing to leave this as is and move on within #297 - correct?

File: api/src/main/java/io/onetable/client/PerTableConfig.java (hunk 3 above)
Comment (lmccay) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:25]: Interpreting this as leave it as it is and move on within #297. Is that correct, @ashvina ?

**Comment by lmccay** [comment]
@the-other-tim-brown , @ashvina - curious where we are on this. I'd rather we get this in before having to rebase it if things move out from under it.

File: api/src/main/java/io/onetable/client/PerTableConfig.java (hunk 3 above)
Comment (ashvina) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:25]: Hi. I apologize for the late reply. I had a slightly different understanding. The objective is to keep `api` free from table format specific configs. This interface does not address that and does not provide much benefit here. In fact, it causes many minor changes (for example, there are many `PerTableConfig` to `PerTableConfigImpl` changes in test cases).

I think we should remove the interface, just relocate the existing `PerTableConfig`.

Comment (lmccay) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:25]: @ashvina  - This interface is necessary for the ServiceLoader to avoid dependency on implementation classes. We can't move the implementation class there as it will pull in other implementation classes due to how it defaults certain values. In which case, we would need to add additional implementation classes or a dependency on core. This interface only has class (interface and method) names that that are format specific but are not implementation classes which I imagined would be changed as part of the PerTableConfig refactoring.

The other option that I can see is that we put this in as is and we follow up with some change to move the implementation class into api that is decoupled from the format specific implementation classes. I'd actually like to see that refactoring result in quite the opposite and PerTableConfig not be in the api at all with injection removing the need for it in that layer.

Bottom line is that today PerTableConfig is dependent on format specific classes. The intent of this PR is not to fix that but to allow for ServiceLoader to work and one of the goals of #297 should indeed be to avoid those dependencies at least within the api layer.

Comment (ashvina) [reply api/src/main/java/io/onetable/client/PerTableConfig.java:25]: I appreciate your clarification. I have nothing else to do regarding this PR. However, I believe #297 should be our next priority.

**Review by ashvina — APPROVED** [review]
LGTM 👍

**Review by the-other-tim-brown — APPROVED** [review]
LGTM! Time to start the next refactoring :)
